### PR TITLE
Add support for MATLAB files

### LINKED ISSUE #25: Add support for matlab files
## Detailed Description

We have some labchart-exported [matlab files in our OSF](https://osf.io/2j43t/). We could implement an interface function that reads matlab files, in order to support yet another type of data!

## Possible Implementation

- Look at the functions in [`interfaces`](https://github.com/physiopy/phys2bids/tree/master/phys2bids/interfaces)
- Using [scipy](https://docs.scipy.org/doc/scipy/reference/tutorial/io.html) or [h5py](https://www.h5py.org/) (or both), write an itnerface function to read Labchart data in matlab format. 
- Write a test for such interface function!

### PR DESCRIPTION
**Pull request by eurunuela**
Closes #25

## Proposed Changes

  - Create a MATLAB interface to read MATLAB files and transform them into `BlueprintInput` objects.
  - Add unit tests for the MATLAB interface.
  - Add integration test for a MATLAB file.

### CODE REVIEW

**Comment by eurunuela** [comment]
@RayStick it would be great if you could make sure that the inputs to phys2bids in the integration test are correct; i.e. `tr=1.2` and `num_timepoints_expected=100`. 

I think `num_timepoints_expected` is correct but I have no idea about `tr`.

**Comment by tsalo** [comment]
Is this a good chance to talk about the organization of these I/O functions? I find the `phys2bids.interfaces.[modality].populate_phys_input()` structure unnecessarily confusing. Why not have `phys2bids.io.load_matlab()`, `phys2bids.io.load_text()`, and `phys2bids.io.load_acq()`?

**Comment by eurunuela** [comment]
> Is this a good chance to talk about the organization of these I/O functions? I find the `phys2bids.interfaces.[modality].populate_phys_input()` structure unnecessarily confusing. Why not have `phys2bids.io.load_matlab()`, `phys2bids.io.load_text()`, and `phys2bids.io.load_acq()`?

I completely agree and did think of it while working this PR. I think it's a very good point we should bring to our next developer call. However, I don't think that should stop this PR from being merged. Actually, it will probably be easier to work on those changes once this PR is merged.

**Comment by tsalo** [comment]
Thanks @eurunuela. You're right- this function can always be moved/renamed at a later date. There's no need to hold up the PR based on that. I'll just open a separate issue about it.

**Comment by vinferrer** [comment]
> 
> 
> @RayStick it would be great if you could make sure that the inputs to phys2bids in the integration test are correct; i.e. `tr=1.2` and `num_timepoints_expected=100`.
> 
> I think `num_timepoints_expected` is correct but I have no idea about `tr`.

You know, I think you can check that in OSF at least in some of the files

**Comment by vinferrer** [comment]
In any case, seems solid code.

**Comment by eurunuela** [comment]
> You know, I think you can check that in OSF at least in some of the files

I haven't found any json file corresponding to this file.

File: phys2bids/tests/test_integration.py
Comment (vinferrer) [inline phys2bids/tests/test_integration.py:249]: I have checked OSF and there isn't any json for ```test_2minRest_´´ so yeah we need @RayStick to tell us if the ``tr`` and ``num_timepoints_expected`` are correct

**Comment by vinferrer** [comment]
Sry i didn't read you

> 
> 
> > You know, I think you can check that in OSF at least in some of the files
> 
> I haven't found any json file corresponding to this file.

**Comment by RayStick** [comment]
I did not upload the test_2minRest.mat, I think it was @AyyagariA, so it is not easy for me to find the details (possible, but not easy).   

I would recommend using these files for tests:

First, look in test_data 
Second, if you can't find what you need in there try the files in _chicago/all_exports_. In that folder, I took time to export the same file in **loads** of possible formats, including _.mat_. Therefore, the file to use would be _chicago/all_exports/cut_file/CDB.mat._ There is just one json file for all these exports (_all_exports/CDB.json_) - not sure if that is useful but it shows you that the TR is 0.555 and num_tps is 1100. 

If those data files are not good for your tests, for whatever reason, I can try to track down details about test_2minRest.mat.

**Comment by eurunuela** [comment]
Thank you @RayStick I think we can make it work with that data file. I'll push the changes tomorrow and hopefully we can have this merged in by the end of the month 💪🏻

**Comment by eurunuela** [comment]
Just updated the tests to use the file _CBD.mat_ in _labchart/chicago/all_exports_. We should probably move it (or copy it) to the testing folder.

Let me know what you think of this PR @vinferrer @RayStick @smoia @tsalo

File: phys2bids/interfaces/mat.py
Comment (vinferrer) [inline phys2bids/interfaces/mat.py:51]: Here you are assuming t_freq is always in Hz. Are you sure this is the case always?

File: phys2bids/interfaces/mat.py
Comment (vinferrer) [inline phys2bids/interfaces/mat.py:61]: Same here, the sample rate could be in other frequency rather than Hz. I am talking from my experience with txt files.

**Review by vinferrer — CHANGES_REQUESTED** [review]

File: phys2bids/interfaces/mat.py
Comment (eurunuela) [reply phys2bids/interfaces/mat.py:51]: Well, I didn't have any other info in the MATLAB files that we have in OSF. So I had to assume the frequencies are given in Hz (they definitely look like they're in Hz anyway).

Is this always the case @RayStick @smoia @tsalo ?

Comment (RayStick) [reply phys2bids/interfaces/mat.py:51]: I don't know if this is always the case (have not used this .mat formats before, generated from a labchart export) but I would assume it be Hz considering that is the units LabChart uses. I guess a .mat file that is not generated from LabChart could be different

Comment (vinferrer) [reply phys2bids/interfaces/mat.py:51]: Well I suppose that's something we have to specify then

File: phys2bids/interfaces/mat.py
Comment (vinferrer) [inline phys2bids/interfaces/mat.py:19]:
```suggestion
    Populate object phys_input from Labchart files in MATLAB format.

```

File: phys2bids/interfaces/mat.py
Comment (vinferrer) [reply phys2bids/interfaces/mat.py:51]: We should see if for example ``acq`` files can be in matlab format and add it to this function

Comment (vinferrer) [reply phys2bids/interfaces/mat.py:51]: imagen
Seems like acq does have matlab format. Therefore we can either leave it for another PR or add it in this one. In any of those case I would put a condition in the function to check that the ``.mat`` file is only in labchart format

Comment (eurunuela) [reply phys2bids/interfaces/mat.py:51]: What do you mean by "disregarding any other structure"? Do you only want labchart MATLAB files to be supported?

Comment (vinferrer) [reply phys2bids/interfaces/mat.py:51]: So if in this PR you only support the labchart structure, yes i want an if condition that checks something that you think is specific of labchart matlab files. In case you are gonna implement also de acq MATLAB files then accept only those two. Basically i want you to check that the ``.mat`` file people load is not some random MATLAB file

Comment (vinferrer) [reply phys2bids/interfaces/mat.py:51]: Mainly because this way we can raise a error

File: phys2bids/interfaces/mat.py
Comment (smoia) [reply phys2bids/interfaces/mat.py:19]: Actually, I think that the next lines deal with AcqKnowledge import as well?

File: phys2bids/tests/test_integration.py
Comment (smoia) [inline phys2bids/tests/test_integration.py:240]: Asking just because I'm seeing it here: why are we going for this approach when we could just tell pytest to run every test _but_ test_integration.py?

File: setup.cfg
Comment (smoia) [inline setup.cfg:45]: See comment above about possibly adding h5py

File: phys2bids/interfaces/mat.py
Comment (smoia) [inline phys2bids/interfaces/mat.py:43]: Ok I have quite a bad news here.
look at the `loadmat` Notes. Matlab files v7.3 aren't supported by it - instead, you need [h5py](https://www.h5py.org/) to [read those](https://stackoverflow.com/a/19340117). Alternatively you can break the code in case a matlab v7.3 file is given as input saying that we don't support it (yet).
In both cases, I would try this import and if it fails act in consequence (either trying using hdf5, or saying we don't support the format, or (better) both.

File: phys2bids/interfaces/mat.py
Comment (smoia) [reply phys2bids/interfaces/mat.py:19]: Ok, after a private chat with Eneko I got what's happening. Reverting this PR to draft!

**Review by smoia — COMMENTED** [review]
Sorry, I went through this PR thinking it was ready to review. Here's just a comment about `loadmat`.
I'm reverting this in draft!

**Comment by smoia** [comment]
@eurunuela @vinferrer, @62442katieb was asking if she can finish this PR. Are you working on it or not any more?

**Comment by eurunuela** [comment]
I think @vinferrer was going to take care of it in the end.

**Comment by 62442katieb** [comment]
@eurunuela @smoia @vinferrer great! then I won't bother

**Comment by eurunuela** [comment]
Let's make sure @vinferrer has done some work on it. I'm pretty sure he has but still.

**Comment by smoia** [comment]
:rocket: Issue was released in [`2.3.0`](https://github.com/physiopy/phys2bids/releases/tag/2.3.0) :rocket:

**Pull request closed without merging**
